<launch>
  <!-- <env name="ROSCONSOLE_CONFIG_FILE" value="$(find nav_configs)/config/potential_gap_rosconsole.conf"/> -->
  
  <arg name="use_recovery_behaviors" default="true"/>
  <arg name="robot_namespace" default="robot0" />
  <arg name="robot_radius" default="0.20"/>
  <arg name="num_agents" default="0"/>
  <arg name="simulation" default="true"/>
  <arg name="rviz" default="false"/>

  <!--<include file="$(find turtlebot_navigation)/launch/includes/velocity_smoother.launch.xml"/>-->

 <node pkg="tf" type="static_transform_publisher" name="world2kmap" args="0 0 0 0 0 0  map known_map 20" />

  <arg name="vel_topic" default="$(arg robot_namespace)/cmd_vel"/>
  <arg name="holonomic" default="true"/>
  <arg name="projection_operator" default="true"/>
  
  <arg name="global_planning_freq" default="$(optenv GM_PARAM_GLOBAL_PLANNING_FREQ 1)"/>
  <arg name="controller_freq" default="$(optenv GM_PARAM_CONTROLLER_FREQ 5)"/>
  <arg name="planning_inflated" default="$(optenv GM_PARAM_PLANNING_INFLATED false)"/>
  <arg name="far_feasible" default="$(optenv GM_PARAM_FAR_FEASIBLE true)"/>
  <arg name="full_fov" default="true"/>
  <arg name="inf_ratio" value="1.2" unless="$(arg planning_inflated)"/>
  <arg name="inf_ratio" value="0.0" if="$(arg planning_inflated)"/>


  <arg name="debug" default="false"/>

  <arg name="debug_prefix" value="" unless="$(arg debug)"/> <!-- valgrind - -tool=callgrind -->
  <arg name="debug_prefix" value="xterm -e gdb -ex run --args" if="$(arg debug)"/>
    
	<!-- necessary for /move_base/make_plan service to work -->
	<node name="known_map" pkg="map_server" type="map_server" args="$(arg map_file)" required="true">
		<remap from="/map" to="/known_map"/>
		<remap from="/map_metadata" to="/known_map_metadata"/>
		<param name="frame_id" value="known_map"/>
	</node>    

  <group if="$(arg simulation)">
    <node pkg="modify_stdr_scan" type="modify_stdr_scan" name="modify_stdr_scan" output="log" 
            args="$(arg robot_namespace) $(arg robot_radius)"/>
  </group>  

  <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="log" clear_params="true" launch-prefix="$(arg debug_prefix)">
      <param name="base_local_planner" value="potential_gap/PotentialGapPlanner" />
      
      <rosparam file="$(find potential_gap)/config/common_costmap_params.yaml" command="load" ns="global_costmap"/>
      <rosparam file="$(find potential_gap)/config/common_costmap_params.yaml" command="load" ns="local_costmap"/>
      <rosparam file="$(find potential_gap)/config/local_costmap_params.yaml" command="load"/>
      <rosparam file="$(find potential_gap)/config/global_costmap_params.yaml" command="load"/>

      <rosparam file="$(find nav_configs)/config/move_base_params.yaml" command="load"/>
      <rosparam file="$(find nav_configs)/config/global_planner_params.yaml" command="load" ns="GlobalPlanner"/>
      <rosparam file="$(find potential_gap)/config/potentialgap_local_planner_params_holonomic.yaml" command="load"/>

      <param name="recovery_behavior_enabled" value="false" unless="$(arg use_recovery_behaviors)"/>

      <param name="controller_frequency" value="$(arg controller_freq)" />
      <param name="planner_frequency" value="$(arg global_planning_freq)"/>

      <remap from="cmd_vel" to="$(arg vel_topic)"/>
      <remap from="/odom" to="$(arg robot_namespace)/odom"/>

      <param name="PotentialGapPlanner/map_frame_id" value="known_map"/>
      <param name="PotentialGapPlanner/odom_frame_id" value="map_static"/>
      <param name="PotentialGapPlanner/robot_frame_id" value="$(arg robot_namespace)"/>
      <param name="PotentialGapPlanner/sensor_frame_id" value="$(arg robot_namespace)_laser_0"/>

      <param name="PotentialGapPlanner/holonomic" value="$(arg holonomic)"/>
      <param name="PotentialGapPlanner/projection_operator" value="$(arg projection_operator)"/>
      <param name="PotentialGapPlanner/planning_inflated" value="$(arg planning_inflated)"/>
      <param name="PotentialGapPlanner/inf_ratio" value="$(arg inf_ratio)"/>
      <param name="PotentialGapPlanner/full_fov" value="$(arg full_fov)"/>
      <param name="PotentialGapPlanner/far_feasible" value="$(arg far_feasible)"/>

      <!-- remapping for obstacles -->
      <param name="PotentialGapPlanner/num_agents" value="$(arg num_agents)" />
      <param name="global_costmap/obstacle_layer/scan/topic" value="$(arg robot_namespace)/mod_laser_0" />
      <param name="global_costmap/robot_base_frame" value="$(arg robot_namespace)" />

      <param name="local_costmap/obstacle_layer/scan/topic" value="$(arg robot_namespace)/mod_laser_0" />
      <param name="local_costmap/robot_base_frame" value="$(arg robot_namespace)" />

  </node>

  <group if="$(arg rviz)">
    <node type="rviz" name="rviz" pkg="rviz" args="-d $(find potential_gap)/rviz/pgap.rviz" />
  </group>

</launch>
